File: mobile/lib/data/services/controller_visibility.dart
import 'package:shared_preferences/shared_preferences.dart';

/// Which optional inputs are shown on the driving screen. Steering Wheel,
/// Accelerator, and Brake are always shown.
class ControllerVisibility {
  const ControllerVisibility({
    required this.showClutch,
    required this.showDashboard,
  });

  final bool showClutch;
  final bool showDashboard;

  static const String clutchKey = 'wheeldeck.show_clutch';
  static const String dashboardKey = 'wheeldeck.show_dashboard';
  static const String legacyKey = 'wheeldeck.controller_type';

  /// Legacy 5-way values mapped to visibility pairs.
  static const Map<String, (bool, bool)> _legacy = {
    'steeringOnly': (false, false),
    'steering3Pedals': (true, false),
    'steering2Pedals': (false, false),
    'steeringDashboard': (false, true),
    'full': (true, true),
  };

  static const ControllerVisibility fallback =
      ControllerVisibility(showClutch: false, showDashboard: true);

  static Future<ControllerVisibility> load() async {
    final prefs = await SharedPreferences.getInstance();
    final legacy = prefs.getString(legacyKey);
    if (legacy != null) {
      final pair = _legacy[legacy] ?? (false, true);
      final migrated = ControllerVisibility(
        showClutch: pair.$1,
        showDashboard: pair.$2,
      );
      await migrated.save();
      await prefs.remove(legacyKey);
      return migrated;
    }
    final clutch = prefs.getBool(clutchKey);
    final dashboard = prefs.getBool(dashboardKey);
    if (clutch == null && dashboard == null) return fallback;
    return ControllerVisibility(
      showClutch: clutch ?? fallback.showClutch,
      showDashboard: dashboard ?? fallback.showDashboard,
    );
  }

  Future<void> save() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(clutchKey, showClutch);
    await prefs.setBool(dashboardKey, showDashboard);
  }
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/services/ControllerVisibility.kt
package dev.fazrigading.wheeldeck.data.services

/// Which optional inputs are shown on the driving screen. Steering Wheel,
/// Accelerator, and Brake are always shown.
data class ControllerVisibility(
    val showClutch: Boolean,
    val showDashboard: Boolean,
) {
    companion object {
        const val CLUTCH_KEY = "wheeldeck.show_clutch"
        const val DASHBOARD_KEY = "wheeldeck.show_dashboard"
        const val LEGACY_KEY = "wheeldeck.controller_type"

        /// Clutch off, dashboard on — what a fresh install drives.
        val fallback = ControllerVisibility(showClutch = false, showDashboard = true)

        /// Legacy 5-way values mapped to visibility pairs; an unknown value
        /// maps to [fallback].
        private val legacy = mapOf(
            "steeringOnly" to ControllerVisibility(showClutch = false, showDashboard = false),
            "steering3Pedals" to ControllerVisibility(showClutch = true, showDashboard = false),
            "steering2Pedals" to ControllerVisibility(showClutch = false, showDashboard = false),
            "steeringDashboard" to ControllerVisibility(showClutch = false, showDashboard = true),
            "full" to ControllerVisibility(showClutch = true, showDashboard = true),
        )

        fun fromLegacy(value: String) = legacy[value] ?: fallback
    }
}
